File: authapp/connectedapps.go
package main

// Connected applications: see what has access to your coursework, and take it
// back (issue #277).
//
// "Disconnect" has to mean more than forgetting a remembered consent, and it
// takes three writes in one request because no single one of them is enough:
//
//  1. Hydra's consent sessions for (subject, client) are revoked. This kills
//     the refresh token immediately — verified: a refresh straight afterwards
//     comes back invalid_grant — so the application cannot mint itself a new
//     access token.
//  2. A row is recorded in data.mcp_grant_revocation. This is what stops the
//     access token the application is *already holding*. mcpapp verifies those
//     offline against Hydra's JWKS, so nothing about step 1 reaches it; what
//     does reach it is that every few minutes it must exchange that token for
//     a fresh internal credential, and api.issue_user_jwt_for_mcp refuses once
//     this row exists. Without it, disconnect would leave up to an hour of
//     working access behind.
//  3. The same row is the audit trail, which is why it is append-only.
//
// The grace period that remains is the internal credential's own lifetime plus
// mcpapp's exchange cache — minutes, not the hour a bare Hydra revocation
// would leave. docs/hydra.md states it.

import (
	"context"
	"crypto/subtle"
	"encoding/json"
	"fmt"
	"html/template"
	"io"
	"log"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"

	"github.com/alexedwards/scs/v2"
)

const (
	connectedAppsPath = "/auth/connected-apps"
	// sessionKeyConnectedAppsCSRF holds the token minted with the page and
	// required back on the disconnect POST.
	sessionKeyConnectedAppsCSRF = "connected_apps_csrf"
)

// connectedApp is one application a student has authorized, as shown to them.
type connectedApp struct {
	ClientID   string   `json:"client_id"`
	ClientName string   `json:"client_name"`
	ClientURI  string   `json:"client_uri,omitempty"`
	Scopes     []string `json:"scopes"`
	// LastActivity is Hydra's handled_at for the most recent consent, which is
	// the last time the user actually approved something for this client —
	// not the last API call, which Hydra does not track.
	LastActivity string `json:"last_activity,omitempty"`
}

// hydraGrantedConsent is the shape Hydra returns from
// GET /admin/oauth2/auth/sessions/consent. Only the fields shown to the user
// are decoded. (hydraConsentSession is a different thing: the session payload
// authapp *sends* when accepting a consent.)
type hydraGrantedConsent struct {
	GrantScope     []string `json:"grant_scope"`
	HandledAt      string   `json:"handled_at"`
	ConsentRequest struct {
		Subject string `json:"subject"`
		Client  struct {
			ClientID   string `json:"client_id"`
			ClientName string `json:"client_name"`
			ClientURI  string `json:"client_uri"`
		} `json:"client"`
	} `json:"consent_request"`
}

// listConnectedApps asks Hydra what the subject has authorized and folds the
// consent sessions into one entry per client. A client the user consented to
// several times has several sessions; they think of it as one application.
func listConnectedApps(ctx context.Context, admin *hydraAdminClient, subject string) ([]connectedApp, error) {
	var sessions []hydraGrantedConsent
	query := url.Values{}
	query.Set("subject", subject)
	if err := admin.request(ctx, http.MethodGet, "/admin/oauth2/auth/sessions/consent", query, nil, &sessions); err != nil {
		var adminErr *hydraAdminError
		// Hydra answers 404/410 when the subject has never consented to
		// anything, which is an empty list rather than a failure.
		if asHydraAdminError(err, &adminErr) && (adminErr.Status == http.StatusNotFound || adminErr.gone()) {
			return nil, nil
		}
		return nil, err
	}

	byClient := map[string]*connectedApp{}
	for _, session := range sessions {
		client := session.ConsentRequest.Client
		if client.ClientID == "" {
			continue
		}
		entry, seen := byClient[client.ClientID]
		if !seen {
			entry = &connectedApp{
				ClientID:   client.ClientID,
				ClientName: client.ClientName,
				ClientURI:  client.ClientURI,
			}
			byClient[client.ClientID] = entry
		}
		entry.Scopes = mergeScopes(entry.Scopes, session.GrantScope)
		if session.HandledAt > entry.LastActivity {
			// RFC 3339 timestamps sort lexically, so this picks the latest.
			entry.LastActivity = session.HandledAt
		}
	}

	apps := make([]connectedApp, 0, len(byClient))
	for _, entry := range byClient {
		if entry.ClientName == "" {
			entry.ClientName = entry.ClientID
		}
		sort.Strings(entry.Scopes)
		apps = append(apps, *entry)
	}
	sort.Slice(apps, func(i, j int) bool {
		if apps[i].LastActivity != apps[j].LastActivity {
			// Most recently used first: that is what a person is looking for.
			return apps[i].LastActivity > apps[j].LastActivity
		}
		return apps[i].ClientName < apps[j].ClientName
	})
	return apps, nil
}

func mergeScopes(existing []string, incoming []string) []string {
	seen := map[string]bool{}
	for _, scope := range existing {
		seen[scope] = true
	}
	for _, scope := range incoming {
		if scope != "" && !seen[scope] {
			seen[scope] = true
			existing = append(existing, scope)
		}
	}
	return existing
}

// revokeConnectedApp performs the disconnect.
//
// The database row goes first, and the order matters. Each write can fail
// independently, so the question is which half-done state is survivable:
//
//   - Row written, Hydra revocation fails: the access token the application
//     holds is already dead (the mint check sees the row), the application is
//     still listed, and the user can press Disconnect again. Recoverable.
//   - Hydra revoked first, row fails: refresh is dead but the access token
//     keeps working for its remaining hour, and the application has vanished
//     from the list — so the user cannot retry the thing that would stop it.
//     Not recoverable, and the page would have told them they were safe.
//
// So the durable stop is written first and Hydra second, and any failure is
// reported rather than swallowed.
func revokeConnectedApp(ctx context.Context, config oauthFlowConfig, jwtConfig FetchJWTConfig, subject string, app connectedApp) error {
	if err := recordGrantRevocation(ctx, jwtConfig, subject, app); err != nil {
		return err
	}

	query := url.Values{}
	query.Set("subject", subject)
	query.Set("client", app.ClientID)
	query.Set("all", "false")
	query.Set("trigger_backchannel_logout", "false")
	if err := config.Admin.request(ctx, http.MethodDelete, "/admin/oauth2/auth/sessions/consent", query, nil, nil); err != nil {
		var adminErr *hydraAdminError
		// Already gone is the desired end state, not an error to show.
		if !asHydraAdminError(err, &adminErr) || !(adminErr.Status == http.StatusNotFound || adminErr.gone()) {
			return fmt.Errorf("revoking consent at hydra: %w", err)
		}
	}
	return nil
}

// recordGrantRevocation writes the row that stops an access token already in
// the wild. A failure here is reported, because a disconnect that only
// revoked at Hydra leaves working access behind and the user must know.
func recordGrantRevocation(ctx context.Context, jwtConfig FetchJWTConfig, netID string, app connectedApp) error {
	payload, err := json.Marshal(map[string]any{
		"netid":       netID,
		"client_id":   app.ClientID,
		"client_name": app.ClientName,
		"scopes":      strings.Join(app.Scopes, " "),
	})
	if err != nil {
		return fmt.Errorf("encoding the revocation record: %w", err)
	}
	endpoint := postgrestRPCURL(jwtConfig, "record_mcp_grant_revocation")
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, strings.NewReader(string(payload)))
	if err != nil {
		return fmt.Errorf("building the revocation request: %w", err)
	}
	request.Header.Set("Authorization", "Bearer "+jwtConfig.AuthappJWT)
	request.Header.Set("Content-Type", "application/json")
	client := &http.Client{Timeout: 5 * time.Second}
	response, err := client.Do(request)
	if err != nil {
		return fmt.Errorf("recording the revocation: %w", err)
	}
	defer response.Body.Close()
	_, _ = io.Copy(io.Discard, io.LimitReader(response.Body, 1<<16))
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		return fmt.Errorf("recording the revocation: postgrest returned %s", response.Status)
	}
	return nil
}

// getConnectedAppsHandler serves the list and the disconnect action for the
// signed-in user. It is session-authenticated like the rest of /auth: there is
// no bearer-token path, because an application must not be able to disconnect
// another application on the user's behalf.
func getConnectedAppsHandler(config oauthFlowConfig, jwtConfig FetchJWTConfig, sessionManager *scs.SessionManager) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		netID := sessionManager.GetString(r.Context(), "netid")
		if netID == "" {
			http.Error(w, "not signed in", http.StatusUnauthorized)
			return
		}

		switch r.Method {
		case http.MethodGet:
			serveConnectedApps(w, r, config, sessionManager, netID)
		case http.MethodPost:
			serveDisconnect(w, r, config, jwtConfig, sessionManager, netID)
		default:
			w.Header().Set("Allow", "GET, POST")
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		}
	})
}

func serveConnectedApps(w http.ResponseWriter, r *http.Request, config oauthFlowConfig, sessionManager *scs.SessionManager, netID string) {
	apps, err := listConnectedApps(r.Context(), config.Admin, netID)
	if err != nil {
		log.Printf("connected apps: listing for %q failed: %v", netID, err)
		http.Error(w, "could not read your connected applications", http.StatusBadGateway)
		return
	}

	if wantsJSON(r) {
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Cache-Control", "no-store")
		_ = json.NewEncoder(w).Encode(map[string]any{"connected_apps": apps})
		return
	}

	token, err := newCSRFToken()
	if err != nil {
		http.Error(w, "could not prepare the page", http.StatusInternalServerError)
		return
	}
	sessionManager.Put(r.Context(), sessionKeyConnectedAppsCSRF, token)

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	w.Header().Set("Cache-Control", "no-store")
	if err := connectedAppsTemplate.Execute(w, map[string]any{
		"NetID":     netID,
		"Apps":      apps,
		"CSRFToken": token,
		"Notice":    r.URL.Query().Get("notice"),
	}); err != nil {
		log.Printf("connected apps: rendering failed: %v", err)
	}
}

func serveDisconnect(w http.ResponseWriter, r *http.Request, config oauthFlowConfig, jwtConfig FetchJWTConfig, sessionManager *scs.SessionManager, netID string) {
	if err := r.ParseForm(); err != nil {
		http.Error(w, "malformed form", http.StatusBadRequest)
		return
	}

	expected := sessionManager.GetString(r.Context(), sessionKeyConnectedAppsCSRF)
	submitted := r.PostForm.Get("csrf_token")
	if expected == "" || subtle.ConstantTimeCompare([]byte(expected), []byte(submitted)) != 1 {
		http.Error(w, "this form has expired; reload the page and try again", http.StatusForbidden)
		return
	}
	// One token, one disconnect: a replayed form cannot cut off a second
	// application the user did not choose.
	sessionManager.Remove(r.Context(), sessionKeyConnectedAppsCSRF)

	clientID := strings.TrimSpace(r.PostForm.Get("client_id"))
	if clientID == "" {
		http.Error(w, "no application was named", http.StatusBadRequest)
		return
	}

	// Take the details from what the user actually has, so a forged form
	// cannot record a revocation naming an application they never authorized,
	// and cannot disconnect on behalf of someone else.
	apps, err := listConnectedApps(r.Context(), config.Admin, netID)
	if err != nil {
		log.Printf("connected apps: listing for %q failed: %v", netID, err)
		http.Error(w, "could not read your connected applications", http.StatusBadGateway)
		return
	}
	var target *connectedApp
	for i := range apps {
		if apps[i].ClientID == clientID {
			target = &apps[i]
			break
		}
	}
	if target == nil {
		http.Error(w, "that application is not connected to your account", http.StatusNotFound)
		return
	}

	if err := revokeConnectedApp(r.Context(), config, jwtConfig, netID, *target); err != nil {
		log.Printf("connected apps: disconnecting %q for %q failed: %v", clientID, netID, err)
		http.Error(w, "could not fully disconnect that application; try again", http.StatusBadGateway)
		return
	}
	log.Printf("connected_apps_disconnect netid=%q client_id=%q", netID, clientID)

	if wantsJSON(r) {
		w.Header().Set("Content-Type", "application/json")
		w.Header().Set("Cache-Control", "no-store")
		_ = json.NewEncoder(w).Encode(map[string]any{
			"disconnected": clientID,
			"grace_note":   "Access already granted stops within a few minutes; refreshing is refused immediately.",
		})
		return
	}
	http.Redirect(w, r, connectedAppsPath+"?notice="+url.QueryEscape("Disconnected "+target.ClientName), http.StatusSeeOther)
}

func wantsJSON(r *http.Request) bool {
	accept := r.Header.Get("Accept")
	return strings.Contains(accept, "application/json") && !strings.Contains(accept, "text/html")
}

// asHydraAdminError unwraps a hydraAdminError, if that is what this is.
func asHydraAdminError(err error, target **hydraAdminError) bool {
	for err != nil {
		if typed, ok := err.(*hydraAdminError); ok {
			*target = typed
			return true
		}
		unwrapper, ok := err.(interface{ Unwrap() error })
		if !ok {
			return false
		}
		err = unwrapper.Unwrap()
	}
	return false
}

// The page is deliberately JS-free, like the consent page it complements: it
// is reached from a browser session, it does one thing, and it must work in
// whatever the student is using.
var connectedAppsTemplate = template.Must(template.New("connected-apps").Funcs(template.FuncMap{
	"friendlyTime": func(value string) string {
		parsed, err := time.Parse(time.RFC3339, value)
		if err != nil {
			return value
		}
		return parsed.Format("2 January 2006, 15:04 MST")
	},
}).Parse(`<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Connected applications</title>
<link rel="stylesheet" href="/auth/oauth/consent.css">
</head>
<body>
<main class="card">
<h1>Connected applications</h1>
<p class="subtitle">Signed in as {{.NetID}}</p>
{{if .Notice}}<p class="notice">{{.Notice}}</p>{{end}}
{{if .Apps}}
<p>These applications can reach your course data using your account. Disconnect
anything you do not recognise or no longer use.</p>
<ul class="apps">
{{range .Apps}}
<li class="app">
  <div class="app-name">{{.ClientName}}</div>
  {{if .ClientURI}}<div class="app-uri">{{.ClientURI}}</div>{{end}}
  <div class="app-scopes">Can: {{range $index, $scope := .Scopes}}{{if $index}}, {{end}}{{$scope}}{{end}}</div>
  {{if .LastActivity}}<div class="app-when">Last approved {{friendlyTime .LastActivity}}</div>{{end}}
  <form method="post" action="` + connectedAppsPath + `">
    <input type="hidden" name="csrf_token" value="{{$.CSRFToken}}">
    <input type="hidden" name="client_id" value="{{.ClientID}}">
    <button type="submit" class="deny">Disconnect</button>
  </form>
</li>
{{end}}
</ul>
<p class="fine-print">Disconnecting stops the application from getting new
access straight away. Access it already holds stops working within a few
minutes.</p>
{{else}}
<p>No applications are connected to your account.</p>
{{end}}
</main>
</body>
</html>
`))

File: tests/oauth/connected-apps.js
/* global describe it before after */

// Group 8: connected applications — seeing what has access, and taking it back
// (issue #277).
//
// The thing worth testing here is that "disconnect" is not cosmetic. Revoking
// consent at Hydra kills the refresh token at once, but the access token the
// application is already holding keeps verifying offline against the JWKS, so
// on its own that leaves up to an hour of working access behind. What closes
// the gap is that mcpapp must exchange that token for a short internal
// credential every ten minutes, and the database refuses once a disconnect is
// on record.
//
// So the tests below assert both halves and the seam between them: refresh
// dies immediately, the already-issued token stops working once the cached
// internal credential is gone, and — the part that makes this usable rather
// than a trap — reconnecting works.
//
// mcpapp is restarted where a test needs the cached credential gone. That is
// standing in for the clock: the cache entry expires with the internal token
// it holds, inside ten minutes, and waiting that out in a test would be silly.

const { expect } = require('chai');
const bunTest = require('bun:test');

const {
    NETIDS,
    READ_SCOPES,
    baseURL,
    sharedClient,
    authorize,
    exchangeCode,
    refreshToken,
    jarFetch,
    McpClient,
    compose,
    sql,
} = require('./helpers.js');

bunTest.setDefaultTimeout(240_000);

const connectedAppsURL = `${baseURL}/auth/connected-apps`;

/** Authorizes the shared client and returns {tokens, jar}. */
async function connect(netid = NETIDS.student, scope = READ_SCOPES) {
    const client = await sharedClient();
    const auth = await authorize({ clientId: client.client_id, netid, scope });
    const exchanged = await exchangeCode({
        code: auth.code, verifier: auth.verifier, clientId: client.client_id,
    });
    return { tokens: exchanged.body, jar: auth.jar, clientId: client.client_id };
}

/** Reads the page and returns its CSRF token. */
async function csrfToken(jar) {
    const response = await jarFetch(connectedAppsURL, {}, jar);
    const html = await response.text();
    const match = html.match(/name="csrf_token" value="([^"]+)"/);
    expect(match, 'the page carried no CSRF token').to.not.equal(null);
    return match[1];
}

async function disconnect(jar, clientId, token) {
    return jarFetch(connectedAppsURL, {
        method: 'POST',
        headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
        body: new URLSearchParams({
            csrf_token: token, client_id: clientId,
        }).toString(),
        redirect: 'manual',
    }, jar);
}

/**
 * Drops mcpapp's exchange cache, standing in for the ten minutes the cached
 * internal credential would otherwise take to expire on its own.
 */
function expireCachedCredentials() {
    compose(['restart', 'mcpapp']);
}

async function callWhoami(accessToken) {
    const mcp = new McpClient(accessToken);
    await mcp.initialize();
    const response = await mcp.callRaw('whoami');
    const result = response.message && response.message.result;
    return {
        isError: Boolean(result && result.isError),
        text: result ? (result.content || []).map(part => part.text || '').join(' ') : '',
        netid: result && result.structuredContent ? result.structuredContent.netid : null,
    };
}

function revocationRows(clientId) {
    const rows = sql(
        'select client_id from data.mcp_grant_revocation '
        + `where client_id = '${clientId}' order by id`,
    );
    return rows === '' ? [] : rows.split('\n');
}

describe('connected applications', () => {
    after(() => {
        // Nothing to undo, and that is worth stating: the rows this file
        // leaves behind are append-only by design and cannot be deleted, but
        // they also cannot poison the rest of the suite. A revocation only
        // refuses tokens issued before it, so every later file's fresh
        // authorization mints normally. The cache is dropped so no credential
        // issued here outlives the file.
        expireCachedCredentials();
    });

    describe('listing', () => {
        it('should refuse an unauthenticated caller', async () => {
            const response = await fetch(connectedAppsURL);
            expect(response.status)
                .to.equal(401);
        });

        it('should show the application a student authorized', async () => {
            const session = await connect();
            const response = await jarFetch(connectedAppsURL, {
                headers: { Accept: 'application/json' },
            }, session.jar);
            expect(response.status)
                .to.equal(200);
            const body = await response.json();
            const app = (body.connected_apps || [])
                .find(entry => entry.client_id === session.clientId);
            expect(app, 'the authorized client was not listed')
                .to.not.equal(undefined);
            // The person needs to recognise it and know what it can reach.
            expect(app.client_name)
                .to.be.a('string')
                .with.length.greaterThan(0);
            expect(app.scopes)
                .to.contain('course:read');
            expect(app.last_activity)
                .to.be.a('string');
        });
    });

    describe('disconnecting', () => {
        it('should refuse a POST with a bad CSRF token, and revoke nothing', async () => {
            const session = await connect();
            const before = revocationRows(session.clientId);
            const response = await disconnect(session.jar, session.clientId, 'not-the-token');
            expect(response.status)
                .to.equal(403);
            expect(revocationRows(session.clientId))
                .to.deep.equal(before);
        });

        it('should refuse to disconnect an application the caller has not authorized', async () => {
            // Otherwise a forged form could record a revocation naming any
            // client at all, and the details written to the audit trail would
            // be whatever the attacker typed.
            const session = await connect();
            const token = await csrfToken(session.jar);
            const response = await disconnect(session.jar, 'some-client-that-is-not-connected', token);
            expect(response.status)
                .to.equal(404);
        });

        it('should sever access and leave an audit row', async () => {
            const session = await connect();

            // It works before, or the rest of this proves nothing.
            const before = await callWhoami(session.tokens.access_token);
            expect(before.netid)
                .to.equal(NETIDS.student);

            const token = await csrfToken(session.jar);
            const response = await disconnect(session.jar, session.clientId, token);
            expect(response.status)
                .to.equal(303);

            // The audit row, carrying what the grant could reach when it was cut.
            const recorded = sql(
                'select netid, client_id, scopes from data.mcp_grant_revocation '
                + `where client_id = '${session.clientId}' order by id desc limit 1`,
            );
            expect(recorded)
                .to.contain(NETIDS.student);
            expect(recorded)
                .to.contain('course:read');

            // Refresh dies immediately: the application cannot mint itself a
            // new access token.
            const refreshed = await refreshToken({
                refresh_token: session.tokens.refresh_token,
                clientId: session.clientId,
            });
            expect(refreshed.status)
                .to.equal(400);
            expect(refreshed.body.error)
                .to.equal('invalid_grant');

            // The access token it already holds is still a valid signature, so
            // this is the half that needs the database: once the cached
            // internal credential is gone, no new one is issued.
            expireCachedCredentials();
            const after = await callWhoami(session.tokens.access_token);
            expect(after.isError, 'a disconnected application still reached course data')
                .to.equal(true);
            expect(after.text)
                .to.contain('disconnected');
        });

        it('should let the student reconnect afterwards', async () => {
            // Without this, disconnecting would be a one-way door: the record
            // is permanent, so the mint check compares it against the token's
            // issued-at rather than merely asking whether a record exists.
            const session = await connect();
            const token = await csrfToken(session.jar);
            expect((await disconnect(session.jar, session.clientId, token)).status)
                .to.equal(303);
            expireCachedCredentials();
            expect((await callWhoami(session.tokens.access_token)).isError)
                .to.equal(true);

            // A fresh authorization: a new grant, and a token issued after the
            // revocation. The pause is not incidental — OAuth's iat has
            // one-second granularity, and the check deliberately treats a
            // token issued in the same second as the revocation as revoked,
            // so a reconnection has to land in a later second. A person
            // clicking through CAS and a consent screen always does; this
            // test is faster than a person.
            // Long enough to clear the five-second clock-skew allowance the
            // mint check applies (see api.issue_user_jwt_for_mcp). A person
            // clicking through CAS and consent always takes longer.
            await new Promise((resolve) => { setTimeout(resolve, 6000); });
            const again = await connect();
            expireCachedCredentials();
            const after = await callWhoami(again.tokens.access_token);
            expect(after.isError, 'a reconnected application was still refused')
                .to.equal(false);
            expect(after.netid)
                .to.equal(NETIDS.student);
        });
    });
});
